tool/vapid.c

```diff
@@ -0,0 +1,381 @@
+      result += vapid_json_escape_unicode(str[i], result);
+    } else {
+      *result++ = str[i];
+    }
+  }
+  *result++ = '"';
+  quotedStr[quotedLen] = '\0';
+  return quotedStr;
+}
+
+// Builds and returns the signature base string. This is what we'll sign with
+// our private key. The base string is *not* null-terminated.
+static char*
+vapid_build_signature_base(const char* aud, size_t audLen, uint32_t exp,
+                           const char* sub, size_t subLen, size_t* sigBaseLen) {
+  char* quotedAud = NULL;
+  char* quotedSub = NULL;
+  char* payload = NULL;
+  char* sigBase = NULL;
+
+  // Build the payload, which contains the audience, expiry, and subject claims.
+  // Since we only need to include three claims, and since this tool is meant to
+  // show how Vapid works with few dependencies, we build our JSON string using
+  // `sprintf`. I don't recommend this approach; it's almost always better to
+  // use a proper serialization library.
+  quotedAud = vapid_json_quote(aud, audLen);
+  if (!quotedAud) {
+    goto end;
+  }
+  quotedSub = vapid_json_quote(sub, subLen);
+  if (!quotedSub) {
+    goto end;
+  }
+  int payloadLen =
+    snprintf(NULL, 0, "{\"aud\":%s,\"exp\":%" PRIu32 ",\"sub\":%s}", quotedAud,
+             exp, quotedSub);
+  if (payloadLen <= 0) {
+    goto end;
+  }
+  // Allocate an extra byte for the null terminator, which `sprintf` appends.
+  payload = malloc(payloadLen + 1);
+  if (!payload) {
+    goto end;
+  }
+  if (sprintf(payload, "{\"aud\":%s,\"exp\":%" PRIu32 ",\"sub\":%s}", quotedAud,
+              exp, quotedSub) <= 0) {
+    goto end;
+  }
+
+  // Determine the Base64url-encoded sizes of the header and payload, and
+  // allocate a buffer large enough to hold the encoded strings and a `.`
+  // separator.
+  size_t encodedHeaderLen =
+    ece_base64url_encode((const uint8_t*) VAPID_HEADER, VAPID_HEADER_LENGTH,
+                         ECE_BASE64URL_OMIT_PADDING, NULL, 0);
+  size_t encodedPayloadLen = ece_base64url_encode(
+    (const uint8_t*) payload, payloadLen, ECE_BASE64URL_OMIT_PADDING, NULL, 0);
+  *sigBaseLen = encodedHeaderLen + encodedPayloadLen + 1;
+  sigBase = malloc(*sigBaseLen);
+  if (!sigBase) {
+    goto end;
+  }
+
+  // Finally, write the encoded header, a `.`, and the encoded payload.
+  ece_base64url_encode((const uint8_t*) VAPID_HEADER, VAPID_HEADER_LENGTH,
+                       ECE_BASE64URL_OMIT_PADDING, sigBase, encodedHeaderLen);
+  sigBase[encodedHeaderLen] = '.';
+  ece_base64url_encode((const uint8_t*) payload, payloadLen,
+                       ECE_BASE64URL_OMIT_PADDING,
+                       &sigBase[encodedHeaderLen + 1], encodedPayloadLen);
+
+end:
+  free(quotedAud);
+  free(quotedSub);
+  free(payload);
+  return sigBase;
+}
+
+// Signs a signature base string with the given `key`, and returns the raw
+// signature.
+static uint8_t*
+vapid_sign(EC_KEY* key, const void* sigBase, size_t sigBaseLen,
```

Comment:
This was the interesting part. OpenSSL's higher-level `EVP_DigestSign*` API produces a DER-encoded signature, but [RFC 7518, section 3.4](https://tools.ietf.org/html/rfc7518#section-3.4) says we should include R and S (in that order).

The Node JWS library uses [`ecdsa-sig-formatter`](https://www.npmjs.com/package/ecdsa-sig-formatter) to extract R and S from the DER signature. I think the Python implementation will need similar treatment.